### PR TITLE
feat: add Antigravity auth slash command for terminal surfaces

### PR DESCRIPTION
**Pull request by zhangyang-crazy-one**
## What

Registers an `antigravity-auth` slash command for DSH surfaces that host the `commands` seam (for example the `deepseek-tui` profile), giving a terminal the same login lifecycle the Web settings card provides:

```text
/antigravity-auth            # show current login state (default)
/antigravity-auth login      # open Google sign-in in the browser and start the OAuth flow
/antigravity-auth cancel     # cancel a pending authorization
/antigravity-auth logout     # clear the shared Antigravity credential
```

The command runs Host-side against the shared `AntigravityAuthService`. `login` acknowledges the unofficial-channel risk note, starts the loopback OAuth flow (its temporary callback listener binds `127.0.0.1:51121`, independent of any DSH WebServer), and opens the Google sign-in page in the default browser with a best-effort platform opener — the authorization URL is never echoed into the command result, because `CommandResult.text` is persisted verbatim into the session's `command/done` event.

## Why

Today the only entry point into the OAuth flow is the Web settings card, which is unavailable on terminal profiles.

## Account-control policy

Account operations are the local-terminal login entry point. The command is enabled when the Host composes **no WebServer** (only local UI adapters can dispatch) or a **loopback-bound WebServer**, and is denied before touching the auth service when a WebServer exposes the shared `commands` seam on any other interface — DSH 0.1.2-alpha.5 command invocations carry no caller/carrier authority, so those compositions fail closed for every caller. The account RPC keeps its own stricter ADR-0008 loopback-only guard.

## Windows browser handoff

The opener is `open` on macOS, `xdg-open` elsewhere, and `cmd /c start "" "<url>"` on Windows. The Windows URL is quoted and spawned with `windowsVerbatimArguments`, because `cmd` splits an unquoted `&` into a second command and libuv does not quote an argument without whitespace; the empty quoted token keeps `start` from reading the URL as the window title. Every `&`-separated OAuth parameter therefore reaches the browser intact. When the opener cannot start, the command reports that failure without reproducing the URL.

## Verification

- `pnpm run check`: peers, oxlint (0 warnings), host/client typechecks, full vitest (**253 tests / 28 files**), build, package smoke, publint.
- `commandAccountMode` unit cases (no WebServer / loopback / public bind / unknown bind).
- `tests/open-authorization-url.spec.ts`: opener spec and exact spawn argv for a realistic multi-parameter OAuth URL, including the Windows quoted token and `windowsVerbatimArguments`; rejected URLs, opener `error`, and a synchronous spawn throw.
- Handler tests assert the authorization URL reaches the opener and never the result text, and that a failed handoff is reported without the URL.
- Lifecycle boundary covers loopback, public-bind (denied), connection-only, and no-connection compositions.
- `tests/terminal-composition.integration.spec.ts`: on the no-WebServer/connection terminal composition, `status` and `cancel` dispatch through the real DSH command runtime into the shared auth service; `login` (shared service `startLogin()` mocked, actual opener with a fake spawn) delivers the full synthetic URL to the opener while `command/run`/`command/done` retain neither the URL nor its `state`/PKCE challenge.

## Limitation

DSH exposes no public transient-presentation or browser-launch API for plugins, so the login handoff is a best-effort platform opener; a Host without a desktop browser cannot complete interactive Google sign-in from the terminal. That gap is named in the README for a separately scoped upstream task.

## Docs

- `README.md` / `README.zh.md`: "Terminal login command" section reworded for the supported surfaces, the Windows quoting, and the failed-handoff report.
- `CHANGELOG.md`: [Unreleased] entry.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -62,6 +64,7 @@ export function apply(ctx: Context): void {
     cleanup: unprovide,
     label: 'antigravity-auth: OAuth and LLM operations',
   })
+  ctx.inject(['commands'], commandCtx => commandCtx.commands.register(createAntigravityAuthCommand(service)))
```
Comment (suntianc) [inline src/index.ts]: [P1] Apply the account-control access policy to the command entry point

The `commands` service is shared by terminal and Web surfaces, and alpha.5 exposes its execution method remotely. Registering this command whenever that service exists therefore also exposes `status`, `login`, `cancel`, and `logout` to Web session-command callers. These handlers invoke the auth service directly, bypassing the loopback guard on the account RPC.

On a Host bound to `0.0.0.0`, the account RPC returns `loopback-required`, but `/antigravity-auth logout` still clears the Host's shared credential and capability evidence. I reproduced the dispatch bypass with the actual `apply()` and real `CommandRuntime`: the RPC was denied without calling the mocked service, while the command succeeded and called `logout()`.

Please enforce an equivalent Host-side authorization policy here and add regression coverage showing that denied Web compositions cannot reach the auth service through commands. A terminal-specific allowance needs a trustworthy Host-side local-authority fact; the presence of `commands` alone does not establish one.

File: src/auth-command.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        try {
+          const current = await service.status()
+          if (current.login.phase === 'pending') {
+            return { kind: 'error', text: 'an Antigravity authorization is already pending; finish it in the browser or run /antigravity-auth cancel' }
+          }
+          await service.acknowledgeRisk()
+          const started = await service.startLogin()
+          return {
+            kind: 'success',
+            text: `Antigravity authorization started (unofficial Antigravity channel, personal use). Open ${started.authorizationUrl} in your browser; after authorizing, run /antigravity-auth status.`,
```
Comment (suntianc) [inline src/auth-command.ts]: [P2] Keep the OAuth authorization URL out of persisted command results

`CommandResult.text` is not transient terminal-only output. DSH's `CommandRuntime.execute()` copies it verbatim into the session's `command/done` event, which follows normal session persistence. Interpolating `started.authorizationUrl` here therefore records the complete authorization URL, including the OAuth state handle and PKCE challenge, in durable session history rather than keeping it within the temporary login interaction.

I reproduced this with the real command runtime and a synthetic authorization URL: both `state` and `code_challenge` appeared in the appended `command/done` event. This does not expose a token or PKCE verifier, but it broadens the retention and visibility of login-flow data beyond the intended ephemeral OAuth interaction.

Please deliver the authorization link through a supported transient presentation channel and return only ordinary status text through the command result. `recordInput: false` would suppress input recording only; it would not prevent this result text from being persisted. Add coverage that inspects the emitted session events, not just the handler's return value. If the terminal surface lacks such a public presentation seam, document that limitation rather than persisting the URL.

**Review by suntianc — CHANGES_REQUESTED** [review]
Please address the two inline findings before merging: the shared command entry point bypasses the account-control loopback policy, and the login result writes the full OAuth authorization URL into session history.

Validation on an isolated snapshot of 9b74413: offline installation with the frozen lockfile and the full `pnpm run check` passed (222 tests, peers, lint, host/client typechecking, build, package smoke, and publint). An additional temporary reproduction using the actual plugin `apply()`, the real DSH `CommandRuntime`, and a mocked auth service confirmed both the logout bypass and the authorization URL appearing in `command/done` events. The 9 command tests plus that reproduction passed. No real OAuth flow or private endpoint request was executed.

**Comment by zhangyang-crazy-one** [comment]
Addressed both review findings (aa5a1c4).

**[P1] Account-control policy on the command entry point.** `loopback-rpc.ts` now exports `loopbackMode(webServerHost)` (enabled only for an explicit `127.0.0.1` bind, blocked otherwise — same semantics as before, extracted so the RPC guard and the command share one decision). `index.ts` records the mode from the connection inject and hands it to the command factory; when blocked, every operation (`status`, `login`, `cancel`, `logout`) returns the `loopback-required` denial without touching the auth service. Note that per the ADR-0008 semantics a terminal-only profile with no WebServer stays blocked; the command is active on loopback-bound compositions.

Regression coverage at the registration/dispatch boundary: `tests/lifecycle.spec.ts` now mounts `apply()` and proves that a `0.0.0.0` composition denies `logout` through the *registered* command while a `127.0.0.1` composition lets `status` through, and `tests/auth-command.spec.ts` denies each operation off-loopback with no service calls.

**[P2] Authorization URL kept out of persisted command results.** `login` no longer interpolates `started.authorizationUrl` into `CommandResult.text` (which `command/done` persists verbatim, state handle + PKCE challenge included). The result is now ordinary status text; the authorization link is presented by the loopback-bound Web settings surface. README/CHANGELOG document this.

`pnpm run check` passes: 229 tests (9 command + 2 new lifecycle boundary cases), peers, oxlint, host/client typecheck, build, package smoke, publint.

**Comment by zhangyang-crazy-one** [comment]
Rebased this branch onto the current upstream `main` (v0.1.4-alpha.6) and re-verified locally. Only doc files overlapped with the upstream release commit; the sole peer change relative to upstream is re-adding `@deepseek-ai/dsh-commands` (now `^0.1.2-alpha.5 || ^0.1.3-alpha.1`, matching the alpha.6 dual-range style) since the `commands` seam is new to this plugin. `pnpm run typecheck` passes and the command/lifecycle suites are green (25 tests). Ready for re-review — @suntianc

**Comment by suntianc** [comment]
Re-reviewed the current head `8aa21e7a6e47e1223f240d4db8128ee2b3d50660` against `main`.

The original non-Web terminal-login goal is still the relevant acceptance criterion. The revisions made in response to our earlier feedback now leave two gaps in that workflow:

1. **A terminal without DSH WebServer/connection cannot execute the account commands.** `accountMode` starts as `blocked` and is enabled only by the loopback-bound Web composition. The command is registered, but `status`, `login`, `cancel`, and `logout` are rejected before reaching the shared service. See [activation](https://github.com/suntianc/dsh-antigravity-auth/blob/8aa21e7a6e47e1223f240d4db8128ee2b3d50660/src/index.ts#L26-L53) and [dispatch](https://github.com/suntianc/dsh-antigravity-auth/blob/8aa21e7a6e47e1223f240d4db8128ee2b3d50660/src/auth-command.ts#L51-L54).

2. **The terminal no longer receives a usable authorization handoff.** The latest handler awaits `service.startLogin()`, discards its returned authorization URL, and returns only “authorization started.” It does not launch a browser. The README now directs the user to the Web settings card, while the PR body still says the command prints the URL. See [the current login handler](https://github.com/suntianc/dsh-antigravity-auth/blob/8aa21e7a6e47e1223f240d4db8128ee2b3d50660/src/auth-command.ts#L63-L74).

Our previous feedback needs this clarification: the static Web RPC policy should not automatically make local terminal login depend on a DSH WebServer. Also, the authorization-URL finding concerned retention in the session's `command/done` event; OAuth state and a PKCE challenge are not an access token or the PKCE verifier. Removing the terminal's only authorization handoff without replacing it leaves the original feature incomplete.

Please define the supported terminal account-control path, preserve appropriate access control on Web/other transports, and provide a usable terminal authorization handoff while addressing session-history retention—for example through a supported Host browser-launch or transient presentation mechanism, if available. The temporary OAuth callback listener on `127.0.0.1:51121` is separate from the DSH WebServer. If a required public presentation/authority API is missing, identify that precise gap for a separately scoped upstream task.

Please add offline integration coverage through the real command runtime for the intended non-Web terminal composition, including service dispatch and delivery of the authorization handoff. Keep credential-bearing values out of persisted command events, and synchronize the PR description and both READMEs with the final behavior. Updating the wording alone would narrow the scope rather than deliver the original non-Web workflow.

For clarity, the additional stale-`accountMode` teardown reproduction is **not an independent security merge blocker**: it demonstrated an in-process policy inconsistency with a mocked connection, not a reachable unauthorized remote invocation after shutdown.

Local verification of this head passed the complete `pnpm run check` on the DSH `0.1.2-alpha.5` graph: **229 tests / 26 files**, peers, lint, Host/client typechecks, build, package smoke, and publint. No live OAuth/private endpoint or terminal-login validation was performed, and the `0.1.3-alpha.1` source graph was not re-run.

**Comment by zhangyang-crazy-one** [comment]
Thanks for the precise re-review. Addressed on head `118b618`.

**Activation is now composition-correct for the terminal entry point.** The account command starts enabled (a terminal profile composes no public WebServer and can only dispatch through local UI adapters) and is blocked only when the connection inject observes a WebServer bound beyond `127.0.0.1` — the case where remote `commands.execute()` callers exist and DSH gives no per-invocation caller/carrier authority to exempt the local terminal (named as the precise upstream API gap). The account RPC keeps its own ADR-0008 loopback-only guard; `loopbackMode` is unchanged.

**Terminal login handoff restored without the retention regression.** `login` now hands `started.authorizationUrl` to the host default browser via a best-effort platform opener (`open`/`xdg-open`/`cmd /c start`, new `src/open-authorization-url.ts` mirroring the DSH TUI opener), and returns only status text — the URL (state handle + PKCE challenge) never reaches `CommandResult.text`/session history. The `127.0.0.1:51121` callback listener is independent of any DSH WebServer, so the flow completes on terminal-only compositions. A Host without a desktop browser cannot complete interactive sign-in from the terminal; that limitation and the missing public transient-presentation/browser-launch API are documented in the READMEs.

**Coverage.** `commandAccountMode` unit cases; handler tests assert the URL reaches the opener and never the result text; lifecycle boundary covers loopback / public-bind-denied / connection-only / no-connection; and a new `tests/terminal-composition.integration.spec.ts` dispatches `status`/`cancel` through the real DSH command runtime on the no-WebServer/connection composition into the shared auth service. Local gates pass: typecheck, lint (0/0), full vitest (239 tests), build, smoke, publint.

PR description and both READMEs are synchronized with the final behavior.

**Comment by suntianc** [comment]
Re-reviewed head `118b61817e60fe718abc706c47e5155d3766873a` against `main`.

The terminal activation correction is working in the tested compositions: commands reach the shared auth service without DSH WebServer/connection, while the public/unknown Web-bind cases remain denied. The account RPC retains its separate loopback guard. The successful login result also no longer contains the authorization URL.

**One remaining P2: preserve the complete OAuth URL in the Windows browser handoff.**

[`openerSpec()`](https://github.com/suntianc/dsh-antigravity-auth/blob/118b61817e60fe718abc706c47e5155d3766873a/src/open-authorization-url.ts#L39-L41) returns `cmd` with `['/c', 'start', '', url]`, and the subsequent `spawn()` uses default Windows argument handling. A normal authorization URL contains multiple query parameters separated by `&`. Libuv does not add quotes merely because an argument contains `&`, and `cmd` interprets an unquoted `&` as a command separator. Consequently, the browser receives only the URL through its first query parameter; the remaining OAuth parameters, including the redirect URI, state and PKCE challenge, are lost from that URL. The command still returns the “opened browser” success message.

This follows from [libuv's argument quoting](https://github.com/libuv/libuv/blob/v1.x/src/win/process.c#L422-L449) and [Microsoft's documented cmd escaping rules](https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/cmd). I observed the exact Windows argv through the real opener with an injected fake spawner; I did **not** execute Windows or a real Google login on this macOS host.

Please fix the Windows handoff so the browser receives the entire URL unchanged, and add coverage for a realistic multi-parameter OAuth URL. Keep the URL out of persisted output when reporting launch failures as well.

Please also add the requested **login integration and event-retention coverage** to the PR. The current integration suite exercises `status` and `cancel`; the login handler test replaces the whole opener with a mock. I added a temporary offline check around the actual plugin `apply()`, real `CommandRuntime`/`SessionStore`, shared auth service with `startLogin()` mocked, and actual opener with fake spawn. It confirms that the synthetic URL reaches the opener and that the recorded session events, including `command/done`, contain neither the URL nor its state/challenge. This is useful permanent regression coverage alongside the Windows-specific test.

Validation of the unmodified PR snapshot passed the complete `pnpm run check` on the DSH `0.1.2-alpha.5` graph: **239 tests / 27 files**, peers, lint, Host/client typechecks, build, package smoke and publint. The supplementary offline integration check passed separately. No real account authorization, private endpoint, or live-profile validation was performed, and the `0.1.3-alpha.1` source graph was not rerun.

**Comment by zhangyang-crazy-one** [comment]
Fixed on head `3b6eadd`.

**Windows handoff.** `openerSpec('win32', url)` now returns `cmd /c start "" "<url>"` and `openAuthorizationUrl` spawns it with `windowsVerbatimArguments: true`, so the quotes reach `cmd` and the URL is no longer split at its first `&`. The empty quoted token keeps `start` from treating the URL as the window title. `tests/open-authorization-url.spec.ts` asserts the spec and the exact spawn argv for a realistic multi-parameter OAuth URL (`client_id`, percent-encoded `redirect_uri`, `scope`, `code_challenge`, `state`), plus rejected URLs, an opener `error`, and a synchronous spawn throw.

**Login integration and event retention.** `tests/terminal-composition.integration.spec.ts` now also dispatches `/antigravity-auth login` through the real `apply()`, `CommandRuntime`, and `SessionStore` on the no-WebServer/connection composition, with the shared service's `startLogin()` mocked and the actual opener driven by a fake spawn. It asserts the full synthetic URL reaches the opener, that `command/run`/`command/done` contain neither the URL nor `state=`/`code_challenge`, and that a failed launch is reported without the URL.

**Failure reporting.** `openAuthorizationUrl` resolves `false` when the URL is not an Antigravity authorization URL, spawn throws, or the opener emits `error`; the command then returns an error result that does not reproduce the URL. This is the launch-failure report you asked to keep value-free.

`pnpm run check` passes: 253 tests / 28 files, peers, oxlint (0/0), host/client typecheck, build, package smoke, publint. READMEs, CHANGELOG, and the PR description are synchronized. No Windows host or live OAuth flow was executed; the Windows argv is verified through the injected spawner.

One upstream note, outside this repository's scope: the DSH TUI opener this module mirrors (`packages/tui/tui-render/src/open-url.ts` in `deepseek-harness`) still passes `['/c', 'start', '', href]` without quoting, so it has the same unquoted-`&` behavior for OSC 8 links. I did not touch DSH core per this repo's scope rules.

**Comment by zhangyang-crazy-one** [comment]
Addressed the abandoned-handoff case on head `b5a1e16`.

`/antigravity-auth login` no longer refuses while an authorization is pending. The command now delegates to `startLogin()`, which already cancels the previous flow (closing its loopback listener and dropping its verifier) before starting a fresh PKCE exchange — the same transition the Web settings card performs through the `login` RPC endpoint. Closing the browser can therefore be retried immediately instead of waiting out the five-minute TTL, and the success text reports that the previous authorization was replaced.

`tests/auth-command.spec.ts` asserts the restart path (acknowledgeRisk and startLogin each called once, the opener receives the new URL, and the URL/state/PKCE challenge never enter the command result). Full offline `pnpm run check` passes.

**Comment by zhangyang-crazy-one** [comment]
Also fixed the missing-capability case after a real browser login on head `d699873`.

The redirect is served by the OAuth flow's loopback listener, which calls `commitCredential` directly and never passes through the public `completeCallback` wrapper that publishes status. The credential and gates landed on disk, but the running Host never re-evaluated the capability lifecycles, so the LLM provider plus Search/Image/Video routes stayed unregistered until the Host restarted — the model catalog showed no Antigravity models. `commitCredential` now publishes the committed credential at its commit point, so those routes register in the running Host.

`tests/auth-service.spec.ts` drives the loopback listener handler directly and asserts a `watchStatus` listener fires after the callback commits. Full offline `pnpm run check` passes.
